aptos-move/aptos-vm/src/move_vm_ext/session.rs

```diff
@@ -368,3 +458,101 @@ impl<'r, 'l> DerefMut for SessionExt<'r, 'l> {
         &mut self.inner
     }
 }
+
+struct WriteOpConverter<'r> {
+    remote: &'r dyn MoveResolverExt,
+    new_slot_metadata: Option<StateValueMetadata>,
+}
+
+impl<'r> WriteOpConverter<'r> {
+    fn convert(
+        &self,
+        state_key: &StateKey,
+        move_storage_op: MoveStorageOp<Vec<u8>>,
+        legacy_creation_as_modification: bool,
+    ) -> Result<WriteOp, VMStatus> {
+        use MoveStorageOp::*;
+        use WriteOp::*;
+
+        let existing_value_opt = self.remote.get_state_value(state_key).map_err(|_| {
+            VMStatus::Error(
+                StatusCode::STORAGE_ERROR,
+                err_msg("Storage read failed when converting change set."),
+            )
+        })?;
+
+        let write_op = match (existing_value_opt, move_storage_op) {
+            (None, Modify(_) | Delete) => {
+                return Err(VMStatus::Error(
+                    StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                    err_msg("When converting write op: updating non-existent value."),
+                ))
+            },
+            (Some(_), New(_)) => {
+                return Err(VMStatus::Error(
+                    StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
```

Comment:
if we want to alert on unknown invariant violation, we can't use it like this. this is possible similarly with parallel execution and duplicate txn